app/gui2/src/components/ColorRing.vue

```diff
@@ -0,0 +1,187 @@
+ *  - Turns allow constants to be expressed as simple numbers, and can be easily converted to the units used by external
+ *    APIs (radians for math, degrees for culori).
+ */
+
+// If the browser doesn't support OKLCH, the gradient will be specified by computing the number of points specified here
+// in OKLCH, converting to sRGB, and interpolating in HSL. This number has been found to be enough to look close to the
+// intended colors, without excessive gradient complexity (which may affect performance).
+const NONNATIVE_OKLCH_INTERPOLATION_STEPS = 300
+
+const FIXED_RANGE_WIDTH = 1 / 16
+
+const selectedColor = defineModel<string | undefined>()
+const props = defineProps<{
+  matchableColors: Set<string>
+}>()
+const emit = defineEmits<{
+  close: []
+}>()
+
+const svgElement = ref<HTMLElement>()
+
+const interaction = injectInteractionHandler()
+
+onMounted(() => {
+  interaction.setCurrent({
+    cancel: () => emit('close'),
+    pointerdown: (e: PointerEvent) => {
+      if (targetIsOutside(e, svgElement.value)) emit('close')
+      return false
+    },
+  })
+})
+
+const mouseSelectedAngle = ref<number>()
+
+const triangleAngle = computed(() => {
+  if (mouseSelectedAngle.value) return mouseSelectedAngle.value
+  if (selectedColor.value) {
+    const color = parseCssColor(selectedColor.value)
+    if (color?.h) return color.h / 360
+  }
+  return undefined
+})
+
+function cssColor(hue: number) {
+  return formatCssColor(ensoColor(hue))
+}
+
+// === Events ===
+
+function eventAngle(event: MouseEvent) {
+  if (!svgElement.value) return 0
+  const origin = Rect.FromDomRect(svgElement.value.getBoundingClientRect()).center()
+  const offset = Vec2.FromXY(event).sub(origin)
+  return Math.atan2(offset.y, offset.x) / (2 * Math.PI) + 0.25
+}
+
+function ringHover(event: MouseEvent) {
+  mouseSelectedAngle.value = eventAngle(event)
+}
+function ringClick(event: MouseEvent) {
+  mouseSelectedAngle.value = eventAngle(event)
+  if (triangleHue.value != null) selectedColor.value = cssColor(triangleHue.value)
+  emit('close')
+}
+
+// === Gradient colors ===
+
+const fixedRanges = computed(() => {
+  const inputHues = new Set<number>()
+  for (const rawColor of props.matchableColors) {
+    if (rawColor === selectedColor.value) continue
+    const color = parseCssColor(rawColor)
+    const hueDeg = color?.h
+    if (hueDeg == null) continue
+    const hue = hueDeg / 360
+    inputHues.add(hue < 0 ? hue + 1 : hue)
+  }
+  return rangesForInputs(inputHues, FIXED_RANGE_WIDTH / 2)
+})
+
+const triangleHue = computed(() => {
+  const target = triangleAngle.value
+  if (target == null) return undefined
+  for (const range of fixedRanges.value) {
+    if (target < range.start) break
+    if (target <= range.end) return range.hue
+  }
+  return target
+})
+
+// === CSS ===
+
+const cssGradient = computed(() => {
+  const points = gradientPoints(
+    fixedRanges.value,
+    browserSupportsOklch ? 2 : NONNATIVE_OKLCH_INTERPOLATION_STEPS,
```

Comment:
Not sure if this is an adequate enough feature check. `browserSupportsOklch` checks if `oklch` is supported as a general color, but not if it is supported in conic gradients specifically. Firefox notably supports the color space, but [doesn't support interpolation methods in the gradient](https://developer.mozilla.org/en-US/docs/Web/CSS/gradient/conic-gradient#browser_compatibility). This needs specialized checks, something like this:
```js
CSS.supports(`background-image: conic-gradient(in oklch increasing hue, red, blue)`)
CSS.supports(`background-image: conic-gradient(in hsl, red, blue)`)
```
Then, there needs to be a separate codepath where `colorInterpolationMethod` is effectively empty when it is not supported.